File: android/core/src/main/java/app/visly/vml/viewimpl/ScrollViewImpl.kt
/**
 * Copyright (c) Visly Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package app.visly.vml.viewimpl

import android.content.Context
import android.view.View
import android.widget.HorizontalScrollView
import android.widget.ScrollView
import app.visly.vml.*

class ScrollViewImpl(ctx: Context): BaseViewImpl<View>(ctx) {
    internal enum class Direction {
        VERTICAL,
        HORIZONTAL,
    }

    internal var direction = Direction.VERTICAL
    internal var contentInset = 0
    internal var content: VMLView? = null

    override fun measure(width: Float?, height: Float?): Size {
        val content = content
        return if (content != null) {
            val size = content.getSize()
            Size(width ?: size.width, height ?: size.height)
        } else {
            Size(width ?: 0f, height ?: 0f)
        }
    }

    override fun setProp(key: String, value: JsonValue) {
        super.setProp(key, value)

        when (key) {
            "direction" -> {
                direction = when (value) {
                    JsonValue.String("horizontal") -> Direction.HORIZONTAL
                    JsonValue.String("vertical") -> Direction.VERTICAL
                    else -> Direction.VERTICAL
                }
            }

            "content-inset" -> {
                contentInset = when (value) {
                    is JsonValue.Object -> value.toDips(ctx).toInt()
                    else -> 0
                }
            }

            "content" -> {
                content = when (value) {
                    is JsonValue.Object -> VMLViewManager.instance.loadJson(ctx, value)
                    else -> null
                }
            }
        }
    }

    override fun createView(): View {
        val view = if (direction == Direction.VERTICAL) {
            ScrollView(ctx)
        } else {
            HorizontalScrollView(ctx)
        }

        view.clipToPadding = false
        view.isHorizontalScrollBarEnabled = false
        view.isVerticalScrollBarEnabled = false
        return view
    }

    override fun bindView(view: View) {
        super.bindView(view)
        val view = if (direction == Direction.VERTICAL) {
            view as ScrollView
        } else {
            view as HorizontalScrollView
        }

        view.setPadding(
                if (direction == Direction.VERTICAL) 0 else contentInset,
                if (direction == Direction.HORIZONTAL) 0 else contentInset,
                if (direction == Direction.VERTICAL) 0 else contentInset,
                if (direction == Direction.HORIZONTAL) 0 else contentInset)

        view.removeAllViews()
        val content = content?.getView(ctx)
        if (content != null) {
            view.addView(content)
        }
    }
}

File: ios/VMLKit/Classes/ScrollViewImpl.swift
/**
 * Copyright (c) Visly Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import UIKit

internal enum Direction {
    case vertical
    case horizontal
}

internal class ScrollViewImpl: BaseViewImpl {
    internal var direction: Direction = .vertical
    internal var contentInset = Float(0)
    internal var content: VMLRoot? = nil
    
    override func measure(width: CGFloat?, height: CGFloat?) -> CGSize {
        if let content = content {
            return content.sizeThatFits(width: width, height: height)
        } else {
            return CGSize(width: width ?? 0, height: height ?? 0)
        }
    }
    
    override func setProp(key: String, value: JsonValue) {
        super.setProp(key: key, value: value)
        
        switch key {
        case "direction":
            switch value {
            case .String(let value) where value == "vertical": self.direction = .vertical
            case .String(let value) where value == "horizontal": self.direction = .horizontal
            default: return self.direction = .vertical
            }
        case "content-inset": self.contentInset = try! value.asObject().asDimension()
        case "content": self.content = VMLViewManager.shared.loadJson(value)
        default: ()
        }
    }
    
    override func createView() -> UIView {
        return UIScrollView()
    }
    
    override func bindView(_ view: UIView) {
        super.bindView(view)
        
        let view = view as! UIScrollView
        view.alwaysBounceVertical = self.direction == .vertical
        view.alwaysBounceHorizontal = self.direction == .horizontal
        
        let inset = CGFloat(self.contentInset)
        if direction == .vertical {
            view.contentInset = UIEdgeInsets(top: inset, left: 0, bottom: inset, right: 0)
            view.contentOffset = CGPoint(x: 0, y: -inset)
        } else {
            view.contentInset = UIEdgeInsets(top: 0, left: inset, bottom: 0, right: inset)
            view.contentOffset = CGPoint(x: -inset, y: 0)
        }
        
        if let content = self.content {
            content.sizeToFit(
                width: self.direction == .vertical ? view.frame.width : nil,
                height: self.direction == .horizontal ? view.frame.height : nil)
            view.addSubview(content.view)
            view.contentSize = content.view.frame.size
        }
    }
}
